### PR TITLE
Fix issues in cmis.get_transceiver_bulk_status

### PR DESCRIPTION
**Pull request by stephenxs**
#### Description

Fix issues in cmis.get_transceiver_bulk_status

1. In case it fails to read EEPROM, either self.get_rx_power() or self.get_tx_power() can be a list of 'N/A'. Need to test it before calling self.mw_to_dbm otherwise exception `TypeError("'<' not supported between instances of 'str' and 'int'")` will be thrown.
    ```
    Exception occured at DomInfoUpdateTask thread due to TypeError("'<' not supported between instances of 'str' and 'int'")
    Traceback (most recent call last):
    File "/usr/local/lib/python3.9/dist-packages/xcvrd/xcvrd.py", line 1696, in run
        self.task_worker()
    File "/usr/local/lib/python3.9/dist-packages/xcvrd/xcvrd.py", line 1681, in task_worker
        post_port_dom_info_to_db(logical_port_name, self.port_mapping, self.xcvr_table_helper.get_dom_tbl(asic_index), self.task_stopping_event, dom_info_cache=dom_info_cache)
    File "/usr/local/lib/python3.9/dist-packages/xcvrd/xcvrd.py", line 503, in post_port_dom_info_to_db
        dom_info_dict = _wrapper_get_transceiver_dom_info(physical_port)
    File "/usr/local/lib/python3.9/dist-packages/xcvrd/xcvrd.py", line 181, in _wrapper_get_transceiver_dom_info
        return platform_chassis.get_sfp(physical_port).get_transceiver_bulk_status()
    File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sonic_xcvr/sfp_optoe_base.py", line 28, in get_transceiver_bulk_status
        return api.get_transceiver_bulk_status() if api is not None else None
    File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sonic_xcvr/api/public/cmis.py", line 211, in get_transceiver_bulk_status
        bulk_status["rx%dpower" % i] = float("{:.3f}".format(self.mw_to_dbm(rx_power[i - 1]))) if self.get_rx_power_support() else 'N/A'
    File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sonic_xcvr/api/xcvr_api.py", line 16, in mw_to_dbm
        elif mW < 0:
    TypeError: '<' not supported between instances of 'str' and 'int'
    Xcvrd: exception found at child thread DomInfoUpdateTask due to TypeError("'<' not supported between instances of 'str' and 'int'")
    ```
3. It should be a valid case for either self.get_rx_power() or self.get_tx_power() to return None. Handle other fields instead of returning None in this case

### CODE REVIEW

**Review by Junchao-Mellanox — DISMISSED** [review]

**Comment by stephenxs** [comment]
It can be cleanly cherry-picked to both 202205 and 202211
202205:
```
commit 16bc0c01a5ac610606c839c1a4bf5204520f4047
Author: Stephen Sun <stephens@nvidia.com>
Date:   Mon Mar 6 04:07:02 2023 +0000

    Fix issue in cmis.get_transceiver_bulk_status

    1. In case it fails to read EEPROM, either self.get_rx_power() or self.get_tx_power() can be a list of 'N/A'. Need to test it before calling self.mw_to_dbm
    2. It should be a valid case for either self.get_rx_power() or self.get_tx_power() to return None. Handle other fields instead of returning None in this case

    Signed-off-by: Stephen Sun <stephens@nvidia.com>

commit 321a8e77131ce567ec015ba26912b4fd1d0b47e1 (origin/202205)
Author: CliveNi <clive.ni@cloudlight.com.hk>
Date:   Fri Sep 23 09:29:13 2022 +0800

    Cdb fw upgrade (#308)

    * [Cloudlight] QSFP-DD FW upgrade doesn't work (#257)

    - Description
```
202211:

```
commit 0430ae52a98d47daa05de347c0510fa3c0966869 (HEAD)
Author: Stephen Sun <stephens@nvidia.com>
Date:   Mon Mar 6 04:07:02 2023 +0000

    Fix issue in cmis.get_transceiver_bulk_status

    1. In case it fails to read EEPROM, either self.get_rx_power() or self.get_tx_power() can be a list of 'N/A'. Need to test it before calling self.mw_to_dbm
    2. It should be a valid case for either self.get_rx_power() or self.get_tx_power() to return None. Handle other fields instead of returning None in this case

    Signed-off-by: Stephen Sun <stephens@nvidia.com>

commit 2dbc0ea7ccecad1677e2f863cadcdfe451d8fd2a (origin/202211)
Author: mihirpat1 <112018033+mihirpat1@users.noreply.github.com>
Date:   Wed Jan 25 09:49:05 2023 -0800

    Change get_tx_bias return type to list (#342)

    * Change get_tx_bias return type to list

    Signed-off-by: Mihir Patel <patelmi@microsoft.com>
```

**Comment by liat-grozovik** [comment]
@prgeor could you please help review or assign someone?

**Comment by prgeor** [comment]
@stephenxs please provide more context in the description. What is failing? Which CLI?

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 1]
```diff
@@ -207,8 +205,8 @@ def get_transceiver_bulk_status(self):
         for i in range(1, self.NUM_CHANNELS + 1):
             bulk_status["tx%ddisable" % i] = tx_disable[i-1] if self.get_tx_disable_support() else 'N/A'
             bulk_status["tx%dbias" % i] = tx_bias[i - 1]
-            bulk_status["rx%dpower" % i] = float("{:.3f}".format(self.mw_to_dbm(rx_power[i - 1]))) if self.get_rx_power_support() else 'N/A'
-            bulk_status["tx%dpower" % i] = float("{:.3f}".format(self.mw_to_dbm(tx_power[i - 1]))) if self.get_tx_power_support() else 'N/A'
+            bulk_status["rx%dpower" % i] = float("{:.3f}".format(self.mw_to_dbm(rx_power[i - 1]))) if self.get_rx_power_support() and rx_power[i - 1] != 'N/A' else 'N/A'
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:208]: why would rx_power be 'N/A'. If its due to due to i2c failure then that should be fixed because we have so many fields to fix. What about tx_disable[]?

**Comment by stephenxs** [comment]
> @stephenxs please provide more context in the description. What is failing? Which CLI? If the i2c read is failing then we should fix the optics or platform issue?

@prgeor Updated in the description.

**Review by keboliu — DISMISSED** [review]

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 1 above)
Comment (prgeor) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:208]: @stephenxs How do I distinguish the RX power = 'N/A' is because of bad optics or due to the optics itself does not support?

Comment (stephenxs) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:208]: If it is caused by bad optics, there will be other errors from drivers. So we do not distinguish it here.
How do you think?

File: tests/sonic_xcvr/test_cmis.py [hunk 2]
```diff
@@ -1266,9 +1266,40 @@ def test_get_transceiver_info(self, mock_response, expected):
...
+                'tx_disabled_channel': 'N/A',
+                'laser_temperature': 40
+            }
+        ),
+        (
+            [
                 ['N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A'],
                 ['N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A'],
-                False, False, False, False, False, False,
+                ['N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A'],
```
Comment (prgeor) [inline tests/sonic_xcvr/test_cmis.py:1296]: Please add a test case where there is aread error and tx_power and rx_power returned is None

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 1 above)
Comment (prgeor) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:208]: @stephenxs get_rx_power() will return 'None' so that also needs to be taken care of here

Comment (stephenxs) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:208]: `get_rx_power()` returns `None` only if `self.get_rx_power_support()` is `False`. We handle that case by checking `self.get_rx_power_support()`, right?
Same as `tx`.

File: tests/sonic_xcvr/test_cmis.py (hunk 2 above)
Comment (stephenxs) [reply tests/sonic_xcvr/test_cmis.py:1296]: Is this what you want?
```
            [
                ['N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A'],
                ['N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A'],
                ['N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A'],
                'N/A',
                50, 3.3,
                ['N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A', 'N/A'],
                None, =========== tx_power
                None, =========== rx_power
                False, False, False, False, False, False,
                {'monitor value': 40},
                None,
            ],
```
https://github.com/sonic-net/sonic-platform-common/pull/351/files/fb813acb2cd07143c74dfc63717b6a8955deb3c1#diff-c4c26365d2facd90024b8631863c6d9756ebede9a3fd781a0a47848901c37096R1269

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 1 above)
Comment (prgeor) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:208]: get_rx_power() can return None on i2c read failure too. could you check

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 3]
```diff
@@ -500,16 +500,18 @@ def get_tx_power(self):
...
+        tx_power_support = self.get_tx_power_support()
+        if not tx_power_support:
+            return tx_power
+
         if tx_power_support:
             tx_power = self.xcvr_eeprom.read(consts.TX_POWER_FIELD)
             if tx_power is not None:
                 tx_power =  [tx_power['OpticalPowerTx%dField' %i] for i in range(1, self.NUM_CHANNELS+1)]
+            else:
+                return None
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:514]: this else is not required? line 512 `tx_power = None` already

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 4]
```diff
@@ -520,16 +522,18 @@ def get_rx_power(self):
...
+        rx_power_support = self.get_rx_power_support()
+        if not rx_power_support:
+            return rx_power
+
         if rx_power_support:
             rx_power = self.xcvr_eeprom.read(consts.RX_POWER_FIELD)
             if rx_power is not None:
                 rx_power = [rx_power['OpticalPowerRx%dField' %i] for i in range(1, self.NUM_CHANNELS+1)]
+            else:
+                return None
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:536]: @stephenxs  we don't need this else? `rx_power=None` already at line 534 so the return at line 538 will indeed return None when 'rx_power' = None

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 3 above)
Comment (stephenxs) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py]: Updated

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 4 above)
Comment (stephenxs) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py]: Updated

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 5]
```diff
@@ -500,12 +500,12 @@ def get_tx_power(self):
...
-        if tx_power_support is None:
-            return None
-
         tx_power = ["N/A" for _ in range(self.NUM_CHANNELS)]
 
+        tx_power_support = self.get_tx_power_support()
+        if not tx_power_support:
+            return tx_power
+
         if tx_power_support:
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:509]: @stephenxs this check at line 509 is not required because of check at line 506?

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 6]
```diff
@@ -520,12 +520,12 @@ def get_rx_power(self):
...
-        if rx_power_support is None:
-            return None
-
         rx_power = ["N/A" for _ in range(self.NUM_CHANNELS)]
 
+        rx_power_support = self.get_rx_power_support()
+        if not rx_power_support:
+            return rx_power
+
         if rx_power_support:
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:529]: this check at line 529 is not required because of the check at line 526?

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@yxieca please cherry pick to 202205

**Comment by prgeor** [comment]
@StormLiangMS please help cherry pick to 202211

**Comment by keboliu** [comment]
commits already included in 202211, so I add the label.
